### DISCUSSION TITLE
Skip a directory without skipping every single file

### DISCUSSION THREAD

**Discussion by limesqueezy** [abraunegg/onedrive#2301] [Q&A] [answered]
I cannot make neither a custom `config` nor a custom `sync_list` skip the whole directory without it unrolling the whole directory tree. Instead it skips every single file of the directory and if a directory has `n` files you will actually get `n` lil _Skipping item - excluded by skip_dir config_ on your terminal, radically slowing the process.

**Comment by abraunegg** [comment]
@arisioz 
A couple of questions:

1. What application version are you using? Output of `onedrive --version` would be helpful
2. Is this being seen when going through items from online, or local items?

> radically slowing the process.

Highly doubtful. The inclusion / exclusion is highly optimised ... outputing log lines will not be radically slowing the process.

**Reply by limesqueezy** [reply]
Thanks for your reply, my version is `v2.4.23-1+np1` and this happens when I try to fetch items from the cloud (first sync). If I was to not include the verbose flag, it simply seems to be stuck to the same place for a long time (left it there for almost an hour, it still did not move). I wasn't particularly referring to printing log lines as the bottleneck here (even though rendering to screen really **is** slowing down any process, usually more than file I/O) but for **probably** not pruning the whole "skipped" directory. Most probably, I have misread something from the `USAGE.md` file and the process won't prune the said directory.

**Comment by abraunegg** [comment]
@arisioz 
> If I was to not include the verbose flag, it simply seems to be stuck to the same place for a long time (left it there for almost an hour, it still did not move).

I guess what needs to be explained to you is *how* the client operates due to the nature of the OneDrive API.

What you are describing is the processing of all the OneDrive bundle lots. When the client prints `Processing XXX OneDrive items ...` it is processing these as provided by the OneDrive API. 

There is zero control as to how to tell OneDrive how many items to send, so they are sent in 200-240 bundle lots. Depending on how many files + folders you have - the number of item sets sent to you can be estimated: total files and folders online / 220 = number of `Processing X OneDrive items ...` you will see. This is *why*, when not using `--verbose` it may feel like the client is not doing anything .. but it actually is processing data in the background.

What is happening here is 100% normal for the application. There is zero issue here.

Now .. *why* do you see `Skipping item - excluded by skip_dir` for items:

1. You are doing 'client side filtering' and OneDrive online has no idea that you are doing client side filtering - the API sends you all your items / objects stored online - which you need to filter and are filtering with your respective configuration. There is zero way around this.
2. Each 'bundle' provided by OneDrive API has to be gone through - thus, each item has to be checked against your client side filtering rules - this is why you see that message printed out.

Most likely this whole query is due not understanding how this client operates. 

If you want to look at where the client is spending all it's time processing, you can add:
```
display_processing_time = "true"
```

To your 'config' file. This will tell you how long the client is spending performing certain actions. That will help inform you on what the client is doing.

You also make no mention of what hardware you are running this on .... so if you have not much CPU ... of course the processing of data is going to be impacted due to your choice of hardware.

**Reply by limesqueezy** [reply]
> You are doing 'client side filtering' and OneDrive online has no idea that you are doing client side filtering - this sends you all items / objects online - which you need to filter and are filtering with your respective configuration. There is zero way around this.

This answers my question. 

In essence the issue here is that if someone wants to exclude, say a dataset of thousands of files they'd have to go through each and every file when syncing. This happens due to not being able to "communicate the exclusion" to OneDrive from this client you have built. If the OneDrive API permitted the pruning of folders it would dramatically accelerate syncing.

I understand that there was an obvious misunderstanding of the whole architecture by my side and not an issue with your client hence asking it on Q&A under Discussions instead of raising an issue. 

Thank you for taking the time to explain the OneDrive API to me :)

**Comment by abraunegg** [comment, marked as answer]
@arisioz 
Please also note that the exhaustive scan is only done one the first sync - to establish a sync state.

Once the sync state is known, the /delta endpoint can be used which reduces the checks and data sent.

The exhaustive scan also performed on:
* Use of --resync ... do not use it unless the client asks for it
* when applicable triggers are met when the application is run
* when using --monitor governed by the full scan parameter
